### ISSUE TITLE
stack overflow

### ISSUE THREAD

**Issue by mccoysc** [automerge/automerge#137]
`doc=automerge.change(doc,function(doc){
    doc.array=new Array(1024*1024);
})`
this big array would cause a stack-overflow for "Maximum call stack size exceeded.
maybe the "recursive calls" in automerge should be modified by "loop call" to avoid the stack overflow

**Comment by ept** [comment]
@mccoysc Thanks for reporting this; however, I don't get this stack overflow problem. Your code example as given throws an exception `TypeError: Unsupported type of value: undefined`, because `new Array(1024*1024)` creates an array with 1m `undefined` values, and Automerge does not permit `undefined` values (it only allows `null`, for parity with JSON).

Even when I change the code to generate an array of non-undefined values, I don't observe a stack overflow. With an array of 100,000 elements I get no error (but it takes ~15 seconds); with an array of 1m elements I get an out-of-memory error in Node (that is, out of **heap** space, not out of stack space).

In summary, I agree that Automerge is slow and uses way too much memory on large arrays, but I can't reproduce the stack overflow, and I can't think of anywhere in the code that performs recursion on arrays. If you can reliably reproduce this, please let me know in which environment you're running (Node? Browser? Electron?).

**Comment by mccoysc** [comment]
I run it in chrome browser version 70
